drivers/virtio: Pass virtio-mmio devices via the cmdline

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`arm64`, `x86_64`]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

A newly introduced config parameter `VIRTIO_MMIO_LINUX_COMPAT_CMDLINE` controls compatibility with Linux. For more details see the description below.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series provides the changes required to pass virtio-mmio devices via the command line. A newly introduced config parameter controls compatibility with Linux. Specifically, when `VIRTIO_MMIO_LINUX_COMPAT_CMDLINE` is selected, the MMIO driver is configured to accept devices in the Linux format, ie:
```    
virtio_mmio.device = <size>@<base>:<irq>[:<id>]
```
    
If this option is not enabled, use the default libukparam format for this library, ie:
```    
libvirtio_mmio.device = <size>@<base>:<irq>[:<id>]
```    
This is useful for VMMs like Firecracker, that automatically inject device descriptors the command line.

To additionally support `x86_64`, various components are updated to make fdt-based discovery optional. These are the platform bus, `virtio-mmio`, and PCI ECAM.

Finally, a generic function `uk_bus_pf_devmap()` to (re)map devices introduced to the platform bus. This allows removing the dependency of declaring devices at the boot pagetables for platforms that support paging (with the exception of early devices), and serves as a placeholder for additional device configuration in security related implementations like the CCA, or SEV. Notice that because of the current boot flow that initializes the `vmmap` API after the probing the interrupt controller - which is required for bringing up secondary cores - it is not possible for `uk_bus_pf_devmap()` to use `uk_vma_map_dma()`, so we need to resort to the low level paging API and map devices 1:1.

GitHub-Depends-On: #1076
GitHub-Depends-On: #1088
GitHub-Depends-On: #1089
GitHub-Depends-On: #1095
GitHub-Depends-On: #1103
GitHub-Depends-On: #1115

michpappas committed:
drivers/virtio/mmio: Support passing devices via the cmdline

This commit adds support for passing virtio-mmio devices via the
command-line.

The probing function is invoked by ukbus at a priority lower than
the the platform bus, to ensure that the virtio devices have been
probed, to safely register with the parent device.

Add a config parameter to control compatibility with Linux.
Specifically, when VIRTIO_MMIO_LINUX_COMPAT_CMDLINE is selected
the MMIO driver is configured to accept devices in the Linux format,
ie:

  virtio_mmio.device = <size>@<base>:<irq>[:<id>]

If this option is not enabled, use the default libukparam
format for this library, ie:

  libvirtio_mmio.device = <size>@<base>:<irq>[:<id>]

This is useful for VMMs like Firecracker, that automatically inject
device descriptors the command line.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/platform: Add size field to struct pf_device

Since it is possible that devices use more than a single page
of MMIO, add a size field to `struct pf_device` to track the
device region size.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/platform: Make fdt discovery optional

Make fdt-based discovery optional to allow the platform bus to be
selected for virtio-mmio on x86_64.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/virtio/mmio: Make fdt discovery optional

Make ftd-based discovery optional, to allow using virtio-mmio
on x86_64.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/pci: Make PCI ECAM conditional to fdt

The PCI ECAM implementation is tightly coupled to fdt. Given that
the platform bus can now be used on x86_64, make PCI ECAM conditional
to the fdt libraries.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/virtio/mmio: Populate pfdev size

Update virtio_mmio_probe_fdt() to populate the device size.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/platform: Add uk_bus_pf_devmap

Add `uk_bus_pf_devmap()` for mapping device regions. Because of the
current boot flow that initializes the vmmap API after the probing
the interrupt controller - which is required for bringing up secondary
cores - it is not possible to use uk_vma_map_dma() here, so we need to
resort to the low level paging API and map devices 1:1.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Enable linux-compatible virtio-mmio cmdline on Firecracker

Firecracker injects virtio-mmio device descriptors to the command line
using the Linux format. Enable VIRTIO_MMIO_LINUX_COMPAT_CMDLINE for
Firecracker.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

jschlumpp commented on drivers/ukbus/platform/platform_bus.c:
```diff
@@ -230,6 +235,47 @@ static int pf_init(struct uk_alloc *a)
 }
 
+#if CONFIG_PAGING
+#include <uk/errptr.h>
+#include <uk/plat/paging.h>
```
Move these includes to the other includes at the top of the file.

michpappas replied:
These were not meant to be there, thanks.

jschlumpp commented on drivers/ukbus/platform/include/uk/bus/platform.h:
```diff
@@ -97,6 +97,17 @@ struct pf_device {
 };
 UK_TAILQ_HEAD(pf_device_list, struct pf_device);
 
+/**
+ * (Re)map device memory
+ *
+ * Maps a region with attributes suitable for device memory access.
+ * If the region is already mapped, it will be remapped unconditionally.
+ *
+ * @param base Base address of the region to map.
+ * @param size Region size. Must be page-aligned.
+ * @return virtual address
+ */
+__vaddr_t uk_bus_pf_devmap(__u64 base, __sz size);
```
The function definition is only available when paging is enabled. Either provide a stub that also works without paging or make this declaration also dependent on paging.

michpappas replied:
Thanks 👍🏼

jschlumpp commented on drivers/virtio/mmio/Config.uk:
```diff
@@ -15,4 +15,27 @@ config LIBVIRTIO_MMIO_FDT
+
+		libvirtio_mmio.device = <size>@<base>:<irq>[:<id>]
+
+
```
Very minor thing: there are two empty lines

michpappas replied:
ACK

jschlumpp commented on drivers/virtio/mmio/virtio_mmio_cmdl.c:
```diff
@@ -0,0 +1,135 @@
+{
+	char *endptr;
+
+	*sz = strtoull(size_str, &endptr, 0);
```
The question is: error checking, yes or no?

michpappas replied:
That's a good question given that proper error checking for `strtoull` is a nightmare. One possible compromise is to check against some arbitrary min / max. Any opinion?

jschlumpp replied:
The fields are not usually set by a user manually, so checking them is not that important.

jschlumpp commented on drivers/virtio/mmio/virtio_mmio_cmdl.c:
```diff
@@ -0,0 +1,135 @@
+
+	*sz = strtoull(size_str, &endptr, 0);
+
+	switch (*endptr) {
+	case 'k':
+	case 'K':
+		*sz *= 1024;
+		++endptr;
+		break;
+	case 'm':
+	case 'M':
+		*sz *= 1024 * 1024;
+		++endptr;
+		break;
+	case 'g':
+	case 'G':
+		*sz *= 1024 * 1024 * 1024;
+		++endptr;
+		break;
+	default:
+		break;
+	}
```
```suggestion
	switch (*endptr) {
	case 'g':
	case 'G':
		*sz *= 1024;
		/* fall through */
	case 'm':
	case 'M':
		*sz *= 1024;
		/* fall through */
	case 'k':
	case 'K':
		*sz *= 1024;
		++endptr;
		break;
	default:
		break;
	}
```
But I guess this is a matter of preference.

michpappas replied:
Yes, I have seen this pattern before. We can do it that way too, either way looks fine to me

michpappas replied:
And by the way, any objections on adding a macro for `__fallthrough` in `essentials.h`? From what I see it is supported by our minimum compiler version of both GCC (`7.x`) and clang (`9.x`).

michpappas committed:
include/essentials: Provide __fallthrough macro

Add a macro for `__attribute__(fallthrough)` to hint the compiler
about fallthrough cases in switch statements. The fallthough
attributed is supported by the minimum compiler version of both GCC
and clang.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/x86: Replace fallthrough comment markers with macro

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Replace fallthrough comment marker with macro

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukdebug: Replace fallthrough comment markers with macro

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
@mschlumpp I updated the PR according to your feedback. I added a few commits to support the `freestanding` attribute and replace the comment-based markers. The item about whether we should do error handling in `strtoull` is still pending.

PS: I didn't rebase yet to make the changes easier to review

michpappas commented:
@mschlumpp I pushed a fix for the `LIBUKLIBPARAM` config option.

jschlumpp reviewed (approved):
Reviewed-by: Marco Schlumpp <marco@unikraft.io>

rares-miculescu reviewed (approved):
Reviewed-by: Rares Miculescu [miculescur@gmail.com](mailto:miculescur@gmail.com)

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
